src/Microsoft.AspNet.OData.Shared/GetNextPageHelper.cs

```diff
@@ -23,35 +23,54 @@ internal static Uri GetNextPageLink(Uri requestUri, IEnumerable<KeyValuePair<str
             StringBuilder queryBuilder = new StringBuilder();
 
             int nextPageSkip = pageSize;
+            
+            //If no value for skiptoken is provided; revert to using skip 
+            bool useSkipToken = true;
+            if (String.IsNullOrWhiteSpace(skipTokenValue))
+            {
+                useSkipToken = false;
+            }
 
             foreach (KeyValuePair<string, string> kvp in queryParameters)
             {
-                string key = kvp.Key;
+                string key = kvp.Key.ToLowerInvariant();
                 string value = kvp.Value;
                 switch (key)
                 {
                     case "$top":
                         int top;
                         if (Int32.TryParse(value, out top))
                         {
-                            // There is no next page if the $top query option's value is less than or equal to the page size.
+                            // There is no next page if the $top query option's value is less than or equal to the page size. You should not call this API if top <= pagesize.
                             Contract.Assert(top > pageSize);
                             // We decrease top by the pageSize because that's the number of results we're returning in the current page
-                            value = (top - pageSize).ToString(CultureInfo.InvariantCulture);
+                            if (top > pageSize)
+                            {
+                                value = (top - pageSize).ToString(CultureInfo.InvariantCulture);
+                            }
                         }
                         break;
                     case "$skip":
-                        int skip;
-                        if (Int32.TryParse(value, out skip))
+                        //Need to increment skip only if we are not using skiptoken 
+                        if (!useSkipToken) 
                         {
-                            // We increase skip by the pageSize because that's the number of results we're returning in the current page
-                            nextPageSkip += skip;
+                            int skip = 0;
+                            if (Int32.TryParse(value, out skip))
+                            {
+                                // We increase skip by the pageSize because that's the number of results we're returning in the current page
+                                nextPageSkip += skip;
+                            }
                         }
                         continue;
                     default:
                         break;
                 }
 
+                if ((key == "$skip" && useSkipToken) || (key == "$skiptoken" && useSkipToken))
```

Comment:
> && useSkipToken [](start = 76, length = 16)

so it's valid to pass $skiptoken in as a query parameter (and it's written with its current value) as long as I don't provide a separate skipTokenValue parameter when calling the GetNextPageLink?

Reply:
Yes, it does sound weird. The reason for this are following -
1. Issue #986 requested a behavior like this.
2. The request for last page of results may have $skiptoken as a query parameter but no skiptoken value, which is a valid case that would require the $skiptoken to be omitted from the url.

Disclaimer for 2nd point - If EnableQuery is used, this code will not be called for the last page. The point is valid only for using the API explicitly.